bubuku/features/swap_partitions.py

```diff
@@ -0,0 +1,176 @@
+
+    @staticmethod
+    def __find_best_swap_candidate(candidates: list, brokers_gap: int, partition_to_swap_size: int) -> TpData:
+        candidates.sort(key=attrgetter("size"), reverse=True)
+        matching_swap_partition = None
+        smallest_new_gap = brokers_gap
+        for tp in candidates:
+            new_gap = abs(brokers_gap - 2 * abs(tp.size - partition_to_swap_size))
+            if new_gap < smallest_new_gap:
+                smallest_new_gap = new_gap
+                matching_swap_partition = tp
+        return matching_swap_partition
+
+    def __create_swap_partitions_json(self, tp1: TpData, br1: str, tp2: TpData, br2: str) -> list:
+        return [
+            (tp1.topic, tp1.partition, self.__replace_broker(tp1.replicas, br1, br2, tp2.replicas[0] == br2)),
+            (tp2.topic, tp2.partition, self.__replace_broker(tp2.replicas, br2, br1, tp1.replicas[0] == br1))
+        ]
+
+    def __replace_broker(self, replicas: list, broker_to_replace: str, replacement: str, was_leader: bool) -> list:
+        rps = [x for x in replicas if x != broker_to_replace]
+        if was_leader:
+            return [replacement] + rps
+        else:
+            return rps + [replacement]
+
+    def __str__(self):
+        return 'SwapPartitions(fat_broker_id: {}, slim_broker_id: {}, gap: {})'.format(
+            self.fat_broker_id, self.slim_broker_id, self.gap)
+
+
+class CheckBrokersDiskImbalance(Check):
+    def __init__(self, zk: BukuExhibitor, broker: BrokerManager, diff_threshold_kb: int):
+        super().__init__(check_interval_s=900)
+        self.zk = zk
+        self.broker = broker
+        self.diff_threshold_kb = diff_threshold_kb
+
+    def check(self):
+        if self.broker.is_running_and_registered():
+            _LOG.info("Starting broker disk imbalance check")
+            try:
+                return self.create_swap_partition_change()
+            except Exception:
+                _LOG.warn("Error occurred when performing disk imbalance check", exc_info=True)
+        return None
+
+    def create_swap_partition_change(self) -> SwapPartitionsChange:
+        size_stats = self.zk.get_disk_stats()
+        if not size_stats or len(size_stats.keys()) == 0:
+            _LOG.info("No size stats available, imbalance check cancelled")
+            return None
+
+        # find the most "slim" and the most "fat" brokers
+        def free_size_getter(tup):
+            return tup[1]["disk"]["free_kb"]
+
+        slim_broker_id = max((item for item in size_stats.items()), key=free_size_getter)[0]
+        fat_broker_id = min((item for item in size_stats.items()), key=free_size_getter)[0]
+        fat_broker_free_kb = size_stats[fat_broker_id]["disk"]["free_kb"]
+        slim_broker_free_kb = size_stats[slim_broker_id]["disk"]["free_kb"]
+
+        # is the gap is big enough to swap partitions?
+        gap = slim_broker_free_kb - fat_broker_free_kb
+        if gap < self.diff_threshold_kb:
```

Comment:
The reason is because it would create instability in certain partitions and not bring any benefits? Maybe we should write the reason as a comment in the code.